### DISCUSSION TITLE
How to connect to specific database/collection from getStaticProps.

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#37195] [Help]
Ok... so I have a code like this:

```
import { ConnectionToDatabase } from "../utils/connection";

export const getStaticProps = async () => {
  const res = await ConnectionToDatabase;
  const data = await res.json();
  return {
    props: {
      data,
    },
    revalidate: 60,
  };
};

const Inventory = ({ data }) => {
```

The page that I am working is this one:
https://github.com/Elindo586/nextprojectb/blob/main/pages/inventory.js

Normally, using regular REACT app  I use Mongoose connections and Schema to create a route, and get the information from the routes.  Each collection would have a single route and I would fetch the information from each route to create individual charts.

What is the best way to tap into each individual collection from MongoDB using Nextjs?   Should I NOT use Mongoose and the Schemas any more?

If I use the MongoDB connection with each individual collection, then, would I need to have one .env per each single collection that I am going to work with? 

Right now I am getting an error, I think it is because I am using Mongoose to connect to a database, but I am not getting into any collection...

The connection file looks like this...
```
const mongoose = require("mongoose");
require("dotenv").config();

mongoose.connect(process.env.MONGODB_URI, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
});

module.exports = mongoose.connection;

```

**Comment by Elindo586** [comment]
I have changed the code to this trying to follow a MongoDB sample.... getting error

```
import { ConnectionToDatabase } from "../utils/connection";

export const getStaticProps = async () => {
  const { db } = await ConnectionToDatabase();
  const data = await db
    .collection("parts")
    .find({})
    .sort({})
    .limit(1000)
    .toArray();
  return {
    props: {
      data: JSON.parse(JSON.stringify(data)),
    },
    revalidate: 60,
  };
};
```

error - pages\inventory.js (9:43) @ getStaticProps
TypeError: (0 , _utils_connection__WEBPACK_IMPORTED_MODULE_5__.ConnectionToDatabase) is not a function
   7 | 
   8 | export const getStaticProps = async () => {
>  9 |   const { db } = await ConnectionToDatabase();
     |                                           ^

**Reply by icyJoseph** [reply]
I guess you just need to await the connection, there's nothing to invoke there.

```ts
  /** Opens Mongoose's default connection to MongoDB, see [connections docs](https://mongoosejs.com/docs/connections.html) */
  export function connect(uri: string, options: ConnectOptions, callback: CallbackWithoutResult): void;
  export function connect(uri: string, callback: CallbackWithoutResult): void;
  export function connect(uri: string, options?: ConnectOptions): Promise<Mongoose>;
```

From those types I can infer that you need to do:

```ts
export const getStaticProps = async () => {
  const { db } = await ConnectionToDatabase; // no function call
  const data = await db
    .collection("parts")
    .find({})
    .sort({})
    .limit(1000)
    .toArray();
  return {
    props: {
      data: JSON.parse(JSON.stringify(data)),
    },
    revalidate: 60,
  };
};
```

> If I use the MongoDB connection with each individual collection, then, would I need to have one .env per each single collection that I am going to work with?

However... I think you might want to do this. Although I am not sure exactly what you mean there.

```ts
import mongoose, { Mongoose } from "mongoose";
// dotenvs are already loaded by next.js from .env* files

const createConnectionManager = () => {
  let current: Mongoose | null = null;

  return {
    connect: async () => {
      if (current) return current;
      // I think the options you are passing are deprecated/removed from mongoose 
      const instance = await mongoose.connect(process.env.MONGODB_URI);
      return instance;
    },
  };
};

export const manager = createConnectionManager();
```

> https://nextjs.org/docs/basic-features/environment-variables

And then use it like:

```ts
import {manager} from '../utils/connection';

export const getStaticProps = async () => {
  const { db } = await manager.connect();

  const data = await db
    .collection("parts")
    .find({})
    .sort({})
    .limit(1000)
    .toArray();
  return {
    props: {
      data: JSON.parse(JSON.stringify(data)), // why?
    },
    revalidate: 60,
  };
};
```

One thing to consider is that, when building Next.js uses workers, so there might be multiple connections going to the DB, this is not something you should think too much about, but the solution above, sort of skips creating more connections if there's no need for them.

You could also put an event listener for when the worker ends (IIRC, https://stackoverflow.com/questions/14031763/doing-a-cleanup-action-just-before-node-js-exits), and then kill the connection in manager, but perhaps mongoose does that already (I hope)?

One last thing, why do you do stringify and then parse?

**Comment by Elindo586** [comment]
First of all... thank you a billion for all your help.


"One last thing, why do you do stringify and then parse?"

The code is taken from this article here:
Example 3: Next.js Static Generation with MongoDB
https://www.mongodb.com/developer/how-to/nextjs-with-mongodb/

^^ However, they skip the section of utils/mongodb...  for the connection.


---------------------------------

Back to the current code ... I was getting an error so I changed it to equal sign like this:

```
const createConnectionManager = () => {
  let current = Mongoose | null;
```

instead of
```
const createConnectionManager = () => {
  let current: Mongoose | null;
```


but now I am getting a different error... at times this error can be an actual bad string, but I think the URI string is good...

error - MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` 
or `mongoose.createConnection()` is a string.

**Reply by icyJoseph** [reply]
> First of all... thank you a billion for all your help.
> 
> "One last thing, why do you do stringify and then parse?"
> 
> The code is taken from this article here: Example 3: Next.js Static Generation with MongoDB https://www.mongodb.com/developer/how-to/nextjs-with-mongodb/
> 
> ^^ However, they skip the section of utils/mongodb... for the connection.
> 
> Back to the current code ... I was getting an error so I changed it to equal sign like this:
> 
> ```
> const createConnectionManager = () => {
>   let current = Mongoose | null;
> ```
> 
> instead of
> 
> ```
> const createConnectionManager = () => {
>   let current: Mongoose | null;
> ```
> 
> but now I am getting a different error... at times this error can be an actual bad string, but I think the URI string is good...
> 
> error - MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string.


I see, the problem is that you are not using TypeScript, and I made a mistake writing the code here.

```js
import mongoose from "mongoose";

const createConnectionManager = () => {
  let current = null;

  return {
    connect: async () => {
      if (current) return current;
      // I think the options you are passing are deprecated/removed from mongoose 
      const instance = await mongoose.connect(process.env.MONGODB_URI);
      return instance;
    },
  };
};

export const manager = createConnectionManager();
```

This is the type free version. When you did `current = Mongoose | null` you made a bitwise OR operation of Mongoose class or null. 

Anyhow... are you proving the `MONGODB_URI` in your env file? do you have one? The `MONGODB_URI` env var won't be available on client side.

And why have you removed this, you just have a connection to the dB, but need to run your query to get data.

```
  const data = await db
    .collection("parts")
    .find({})
    .sort({})
    .limit(1000)
    .toArray();
```

**Comment by Elindo586** [comment]
so... I got it as .env.local  is that correct? 

It looks like this:
`MONGODB_URI=mongodb+srv://elindo:<passwordhere>@cluster0.gcyyo.mongodb.net/test?retryWrites=true&w=majority`

This all section still there, maybe I didn't update the folder ? 

```
import { manager } from "../utils/connection";

export const getStaticProps = async () => {
  const { db } = await manager.connect();

  const data = await db
    .collection("parts")
    .find({})
    .sort({})
    .limit(1000)
    .toArray();
  return {
    props: {
      data: JSON.parse(JSON.stringify(data)), // why?
    },
    revalidate: 60,
  };
};


```
A good link for this file is this one:
https://github.com/Elindo586/nextprojectb/tree/main/pages 


Speaking of the missing code....  say....  I do the connection the database, but how can the connection know I am tapping into a certain collection?  

The above part of the code is coming from an article from MongoDB.... but the actual connection I am using Mongoose....  so, I don't know how exactly  the code gonna know which collection I am trying to tap from the data base......
